src/libstd/io/impls.rs

```diff
@@ -337,14 +339,21 @@ impl Write for &mut [u8] {
 impl Write for Vec<u8> {
     #[inline]
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if unlikely(self.try_reserve(buf.len()).is_err()) {
+            return write_vec_oom(self, buf);
+        }
         self.extend_from_slice(buf);
         Ok(buf.len())
     }
 
     #[inline]
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
         let len = bufs.iter().map(|b| b.len()).sum();
```

Comment:
The code here assumes that `.sum()` won't overflow usize. I'm not sure this is an safe assumption, but `BufWriter` didn't handle it properly either, so I assumed it's fine.